src/Lucene.Net/Search/FieldCacheImpl.cs

```diff
@@ -782,42 +784,44 @@ public virtual StringIndex GetStringIndex(IndexReader reader, System.String fiel
         {
             return (StringIndex) caches[typeof(StringIndex)].Get(reader, new Entry(field, (Parser) null), state);
         }
-        
+
         internal sealed class StringIndexCache:Cache
         {
             internal StringIndexCache(FieldCache wrapper):base(wrapper)
             {
             }
-            
+
             protected internal override System.Object CreateValue(IndexReader reader, Entry entryKey, IState state)
             {
                 System.String field = StringHelper.Intern(entryKey.field);
                 int[] retArray = new int[reader.MaxDoc];
                 int[] retArrayOrdered = new int[reader.MaxDoc];
-                for (int i = 0; i < retArrayOrdered.Length; i++)
-                {
-                    retArrayOrdered[i] = -1;
+                for (int i = 0; i < retArrayOrdered.Length; i++)
+                {
+                    retArrayOrdered[i] = -1;
                 }
-                System.String[] mterms = new System.String[reader.MaxDoc + 1];
+
+                var length = reader.MaxDoc + 1;
+                UnmanagedStringArray mterms = new UnmanagedStringArray(reader.MaxDoc + 1);
                 TermDocs termDocs = reader.TermDocs(state);
-                TermEnum termEnum = reader.Terms(new Term(field), state);
+                
+                SegmentTermEnum termEnum = (SegmentTermEnum)reader.Terms(new Term(field), state);
```

Comment:
I'm pretty sure that we are guaranteed that this will always be the case here. 
I assume that all the tests are passing?

Reply:
yes